server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java

```diff
@@ -207,17 +207,24 @@ public void onError(Throwable t) {
           LOGGER.error("[{}] stream closed with error", streamId, t);
         }
 
-        callbacks.onStreamCloseWithError(streamId, defaultTypeUrl, t);
-        responseObserver.onError(Status.fromThrowable(t).asException());
-        cancel();
+        try {
+          callbacks.onStreamCloseWithError(streamId, defaultTypeUrl, t);
+          responseObserver.onError(Status.fromThrowable(t).asException());
```

Comment:
maybe this should go in the finally block as well?

Reply:
Hmm, good question. If `onStreamCloseWithError` throws, which throwable would we use in the `Status.fromThrowable` call? I think it's a little more straightforward to leave it as-is for now.